video/out/vo_drm.c

```diff
@@ -0,0 +1,523 @@
+
+    if (mp_sws_reinit(p->sws) < 0)
+        return 1;
+
+    vo->want_redraw = true;
+    return 0;
+}
+
+static void draw_image(struct vo *vo, mp_image_t *mpi)
+{
+    struct priv *p = vo->priv;
+
+    struct modeset_buf *front_buf = &p->dev->bufs[p->dev->front_buf];
+    struct mp_image img = {0};
+    mp_image_set_params(&img, &p->sws->dst);
+    img.planes[0] = p->buf + p->x * 4;
+    img.stride[0] = front_buf->stride;
+
+    struct mp_rect src_rc = p->src;
+    src_rc.x0 = MP_ALIGN_DOWN(src_rc.x0, mpi->fmt.align_x);
+    src_rc.y0 = MP_ALIGN_DOWN(src_rc.y0, mpi->fmt.align_y);
+    mp_image_crop_rc(mpi, src_rc);
+
+    mp_sws_scale(p->sws, &img, mpi);
+
+    osd_draw_on_image(vo->osd, p->osd, mpi ? mpi->pts : 0, 0, &img);
+
+    memmove(front_buf->map, p->buf, p->device_w * p->device_h * 4);
+}
+
+static void flip_page(struct vo *vo)
+{
+    struct priv *p = vo->priv;
+
+    int ret = drmModeSetCrtc(p->fd, p->dev->crtc,
+                             p->dev->bufs[p->dev->front_buf].fb,
+                             0, 0, &p->dev->conn, 1, &p->dev->mode);
+    if (ret) {
+        MP_WARN(vo, "Cannot flip page for connector\n");
+    } else {
+        p->dev->front_buf ++;
+        p->dev->front_buf %= BUF_COUNT;
+    }
+}
+
+static int preinit(struct vo *vo)
+{
+    struct priv *p = vo->priv;
+    p->dev = NULL;
+    p->fd = 0;
+    p->old_crtc = NULL;
+    p->sws = mp_sws_alloc(vo);
+
+    int ret;
+
+    ret = modeset_open(vo, &p->fd, p->device_path);
+    if (ret)
+        return ret;
+
+    ret = modeset_prepare_dev(vo, p->fd, p->connector_id, &p->dev);
+    if (ret)
+        return ret;
+
+    assert(p->dev != NULL);
+    p->device_w = p->dev->bufs[0].width;
+    p->device_h = p->dev->bufs[0].height;
+
+    p->old_crtc = drmModeGetCrtc(p->fd, p->dev->crtc);
+    ret = drmModeSetCrtc(p->fd, p->dev->crtc,
+                         p->dev->bufs[p->dev->front_buf + BUF_COUNT - 1].fb,
+                         0, 0, &p->dev->conn, 1, &p->dev->mode);
+    if (ret) {
+        char *errstr = mp_strerror(errno);
+        MP_ERR(vo, "Cannot set CRTC for connector %u: %s\n", p->connector_id,
+               errstr);
+    }
+
+    return 0;
+}
+
+static void uninit(struct vo *vo)
+{
+    struct priv *p = vo->priv;
+
+    if (p->dev != NULL) {
+        if (p->old_crtc != NULL) {
+            drmModeSetCrtc(p->fd,
+                           p->old_crtc->crtc_id,
+                           p->old_crtc->buffer_id,
+                           p->old_crtc->x,
+                           p->old_crtc->y,
+                           &p->dev->conn,
+                           1,
+                           &p->dev->mode);
+            drmModeFreeCrtc(p->old_crtc);
+        }
+
+        modeset_destroy_fb(p->fd, &p->dev->bufs[1]);
+        modeset_destroy_fb(p->fd, &p->dev->bufs[0]);
+    }
+
+    talloc_free(p->buf);
+    talloc_free(p->dev);
+}
+
+static int query_format(struct vo *vo, int format)
+{
+    return format == IMGFMT_BGR0;
```

Comment:
It would probably be faster if you allowed any swscale format here (like vo_x11).